orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,501 @@
+  return s.rfind(p, 0) == 0;
+}
+
+bool StringEndsWith(std::string const& s, std::string const& p) {
+  if (p.size() > s.size()) return false;
+  return std::equal(p.rbegin(), p.rend(), s.rbegin());
+}
+
+const std::vector<std::string> ParseStringData(
+    const ONNX_NAMESPACE::TensorProto* tensor_proto) {
+  ORT_ENFORCE(tensor_proto->data_type() == ONNX_NAMESPACE::TensorProto_DataType_STRING,
+              "ParseStringData type mismatch for tensor ");
+
+  ORT_ENFORCE(!(tensor_proto->has_data_location() &&
+                tensor_proto->data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL),
+              "Cannot parse string data from external tensors.");
+
+  ORT_ENFORCE(!tensor_proto->has_raw_data(),
+              "stringcontent is required to be stored in repeated"
+              "bytes string_data field. raw_data type cannot be string.");
+
+  std::vector<std::string> res;
+  const auto& data = tensor_proto->string_data();
+  int expected_size = 1;
+  for (int i = 0; i < tensor_proto->dims_size(); ++i) {
+    expected_size *= tensor_proto->dims(i);
+  }
+
+  ORT_ENFORCE(tensor_proto->dims_size() == 0 || data.size() == expected_size, "Data size mismatch.");
+  res.insert(res.end(), data.begin(), data.end());
+  return res;
+}
+}  // namespace
+
+Status CreateOrtValuesFromTensorProtos(
+    const std::vector<const ONNX_NAMESPACE::TensorProto*>& tensor_protos,
+    NameMLValMap& name_to_ort_value) {
+  static CPUExecutionProviderInfo info;
+  static CPUExecutionProvider cpu_provider(info);
+  static AllocatorPtr cpu_allocator = cpu_provider.GetAllocator(0, OrtMemTypeDefault);
+
+  for (const auto tensor_proto : tensor_protos) {
+    TensorShape tensor_shape{utils::GetTensorShapeFromTensorProto(*tensor_proto)};
+    const DataTypeImpl* tensor_dtype = DataTypeImpl::TensorTypeFromONNXEnum(tensor_proto->data_type())->GetElementType();
+    auto p_tensor = std::make_unique<Tensor>(tensor_dtype, tensor_shape, cpu_allocator);
+    ORT_RETURN_IF_ERROR(utils::TensorProtoToTensor(Env::Default(), nullptr, *tensor_proto, *p_tensor));
+
+    OrtValue ort_value;
+    ort_value.Init(p_tensor.release(), DataTypeImpl::GetType<Tensor>(), DataTypeImpl::GetType<Tensor>()->GetDeleteFunc());
+    name_to_ort_value.emplace(tensor_proto->name(), ort_value);
+  }
+
+  return Status::OK();
+}
+
+Status CheckpointUtils::OrtSaveInternal(
+    const std::string& model_uri,
+    const std::vector<std::string>& trainable_param_names,
+    const PathString& checkpoint_path) {
+  auto logger_ptr = std::make_unique<logging::Logger>(logging::LoggingManager::DefaultLogger());
+  std::shared_ptr<Model> p_model;
+  ORT_RETURN_IF_ERROR(Model::Load(model_uri, p_model, nullptr, *logger_ptr));
+  Graph& graph = p_model->MainGraph();
+
+  std::vector<const ONNX_NAMESPACE::TensorProto*> trainable_weight_values;
+  trainable_weight_values.reserve(trainable_param_names.size());
+  for (size_t i = 0; i < trainable_param_names.size(); ++i) {
```

Comment:
also looks like we are extracting and saving only the trainable params here, different from the design

Reply:
Yeah need fix it. I fixed the issue for loading, missed the saving.